### PR TITLE
feat(TCOMP-2739):Create an action validator for the @DynamicDependencies annotation

### PR DESCRIPTION
**Pull request by yyin-talend**
https://jira.talendforge.org/browse/TCOMP-2739

### Requirements

* Any code change adding any logic **MUST** be tested through a unit test executed with the default build
* Any API addition **MUST** be done with a documentation update if relevant

### CODE REVIEW

File: component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java [hunk 1]
```diff
@@ -206,6 +213,51 @@ private Stream<String> findDiscoverSchemaExtendedErrors(final AnnotationFinder f
...
+                .stream()
+                .filter(m -> !hasListReturnType(m))
+                .map(m -> m + " should return a List")
+                .sorted();
+
+        final Stream<String> returnStringInList = finder
+                .findAnnotatedMethods(DynamicDependencies.class)
+                .stream()
+                .filter(m -> !hasStringInList(m))
+                .map(m -> m + " should return String in the List")
```
Comment (undx) [inline component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java:246]:
```suggestion
                .map(m -> m + " should return List<String>")
```

**Review by undx — COMMENTED** [review]
I think some stuff are missing
The action should certainly be added here also
https://github.com/Talend/component-runtime/blob/5ab148ca77667d77654b5098c34cfc9f8638ecd6/component-tools/src/main/java/org/talend/sdk/component/tools/validator/Validators.java#L182

File: component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (yyin-talend) [reply component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java]: Combined

Comment (yyin-talend) [reply component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java]: Modified

File: component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java [hunk 2]
```diff
@@ -206,6 +213,44 @@ private Stream<String> findDiscoverSchemaExtendedErrors(final AnnotationFinder f
...
+                .filter(m -> !hasStringInList(m))
+                .map(m -> m + " should return List<String>")
+                .sorted();
+
+        final Stream<String> dataset = finder
+                .findAnnotatedMethods(DynamicDependencies.class)
+                .stream()
+                .filter(m -> !hasTypeParameter(m, DataSet.class))
+                .map(m -> m + " should have its Dataset parameter")
+                .sorted();
```
Comment (undx) [inline component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java:247]: Sorry, did not noticed at first review: Also this one can be coupled with `optionParameter`. We need an `@Option`ed dataset...

File: component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java [hunk 3]
```diff
@@ -294,4 +339,20 @@ private boolean hasBranchCorrectNaming(final Method method) {
     private boolean hasCorrectReturnType(final Method method) {
         return Schema.class.isAssignableFrom(method.getReturnType());
     }
+
+    private boolean hasListReturnType(final Method method) {
```
Comment (undx) [inline component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java:343]: Is this method still used, not sure...

**Review by undx — COMMENTED** [review]
See also sonar lint comments and fix them.

File: component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java (hunk 3 above)
Comment (yyin-talend) [reply component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java]: removed

File: component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java [hunk 4]
```diff
@@ -206,6 +213,37 @@ private Stream<String> findDiscoverSchemaExtendedErrors(final AnnotationFinder f
...
+     *
+     * @param finder
+     * @return Errors on @DynamicDependencies method
+     */
+    private Stream<String> findDynamicDependenciesErrors(final AnnotationFinder finder) {
+
+        final Stream<String> optionParameter = finder
+                .findAnnotatedMethods(DynamicDependencies.class)
+                .stream()
+                .filter(m -> !hasOption(m))
```
Comment (undx) [inline component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java:231]: Here, we have to check that `@Option` parameter should be a class defined as a `@DataSet`.

File: component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java [hunk 5]
```diff
@@ -206,6 +213,37 @@ private Stream<String> findDiscoverSchemaExtendedErrors(final AnnotationFinder f
...
+                .map(m -> m + " should have a parameter being an option (marked with @Option)")
```
Comment (undx) [inline component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java:232]:
```suggestion
                .map(m -> m + " should have a DataSet parameter marked with @Option")
```

File: component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java (hunk 4 above)
Comment (yyin-talend) [reply component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java]: Added

File: component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java (hunk 5 above)
Comment (yyin-talend) [reply component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java]: Modified

File: component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java [hunk 6]
```diff
@@ -206,6 +213,37 @@ private Stream<String> findDiscoverSchemaExtendedErrors(final AnnotationFinder f
...
+     *
+     * @param finder
+     * @return Errors on @DynamicDependencies method
+     */
+    private Stream<String> findDynamicDependenciesErrors(final AnnotationFinder finder) {
+
+        final Stream<String> optionParameter = finder
+                .findAnnotatedMethods(DynamicDependencies.class)
+                .stream()
+                .filter(m -> !hasOption(m) || !hasTypeParameter(m, DataSet.class))
```
Comment (undx) [inline component-tools/src/main/java/org/talend/sdk/component/tools/validator/ActionValidator.java:231]: I think that this won't work: `DataSet` is intended to be an annotation to a class, not extending.

File: component-tools/src/test/java/org/talend/sdk/component/tools/validator/ActionValidatorTest.java [hunk 7]
```diff
@@ -263,4 +277,41 @@ public Record guessProcessorSchemaKo6(@Option FakeDataSet configuration, RecordB
             return null;
         }
     }
+
+    @Service
+    static class ActionDynamicDependenciesOK {
+
+        @DynamicDependencies("test-all")
+        public List<String> getDynamicDependencies(@Option("configuration") final DataSet dataset) {
```
Comment (undx) [inline component-tools/src/test/java/org/talend/sdk/component/tools/validator/ActionValidatorTest.java:285]: You're cheating on the test here :-)
`org.talend.sdk.component.api.configuration.type.DataSet` is an annotation ! Same my previous comment.
Instead use the `FakeDataSet` class for accurate test.

File: component-tools/src/test/java/org/talend/sdk/component/tools/validator/ActionValidatorTest.java [hunk 8]
```diff
@@ -263,4 +277,41 @@ public Record guessProcessorSchemaKo6(@Option FakeDataSet configuration, RecordB
...
+            return null;
+        }
+    }
+
+    @Service
+    static class ActionDynamicDependenciesKO {
+
+        @DynamicDependencies("error: return List<String>")
+        public String getDynamicDependencies(@Option("configuration") final DataSet dataset) {
```
Comment (undx) [inline component-tools/src/test/java/org/talend/sdk/component/tools/validator/ActionValidatorTest.java:294]: Same remark as in ` ActionDynamicDependenciesOK`.

File: component-tools/src/test/java/org/talend/sdk/component/tools/validator/ActionValidatorTest.java [hunk 9]
```diff
@@ -263,4 +277,41 @@ public Record guessProcessorSchemaKo6(@Option FakeDataSet configuration, RecordB
...
+        public String getDynamicDependencies3(@Option("configuration") final FakeDataStore dataset) {
+            return null;
+        }
+
+        @DynamicDependencies("error: param not option")
+        public List<Object> getDynamicDependencies4(final FakeDataStore dataset) {return null;
+        }
+
+        @DynamicDependencies("error: List<T> T not String")
+        public List<Object> getDynamicDependencies5(@Option("configuration") final DataSet dataset) {
```
Comment (undx) [inline component-tools/src/test/java/org/talend/sdk/component/tools/validator/ActionValidatorTest.java:313]: Same remark as in ` ActionDynamicDependenciesOK`.

File: component-tools/src/test/java/org/talend/sdk/component/tools/validator/ActionValidatorTest.java [hunk 10]
```diff
@@ -263,4 +277,45 @@ public Record guessProcessorSchemaKo6(@Option FakeDataSet configuration, RecordB
...
+            return null;
+        }
+
+        @DynamicDependencies("error: param not dataset")
+        public String getDynamicDependencies3(@Option("configuration") final FakeDataStore dataset) {
+            return null;
+        }
+
+        @DynamicDependencies("error: param not option")
+        public List<Object> getDynamicDependencies4(final FakeDataStore dataset) {return null;
```
Comment (undx) [inline component-tools/src/test/java/org/talend/sdk/component/tools/validator/ActionValidatorTest.java:309]: formatting

File: component-tools/src/test/java/org/talend/sdk/component/tools/validator/ActionValidatorTest.java [hunk 11]
```diff
@@ -263,4 +277,45 @@ public Record guessProcessorSchemaKo6(@Option FakeDataSet configuration, RecordB
...
+        }
+
+        @DynamicDependencies("error: List<T> T not String")
+        public List<Object> getDynamicDependencies5(@Option("configuration") final FakeDataSet dataset) {
+            return null;
+        }
+
+        @DynamicDependencies("error: 2 params")
+        public List<String> getDynamicDependencies6(@Option("configuration") final FakeDataSet dataset, @Option("configuration") final FakeDataSet dataset2) {return null;
```
Comment (undx) [inline component-tools/src/test/java/org/talend/sdk/component/tools/validator/ActionValidatorTest.java:318]: formatting

**Review by undx — COMMENTED** [review]
2 minors comments

**Review by undx — APPROVED** [review]
lgtm

**Pull request merged**